nexus/db-queries/src/db/datastore/disk.rs

```diff
@@ -73,59 +281,147 @@ impl DataStore {
+                        disk.id(),
+                    )));
+                }
+            }
+        }
+
+        Ok(list)
     }
 
-    pub async fn project_create_disk(
-        &self,
-        opctx: &OpContext,
+    pub(super) async fn project_create_disk_in_txn(
+        conn: &async_bb8_diesel::Connection<DbConnection>,
+        err: OptionalError<Error>,
         authz_project: &authz::Project,
         disk: Disk,
-    ) -> CreateResult<Disk> {
+    ) -> Result<Disk, diesel::result::Error> {
         use nexus_db_schema::schema::disk::dsl;
-
-        opctx.authorize(authz::Action::CreateChild, authz_project).await?;
+        use nexus_db_schema::schema::disk_type_crucible::dsl as disk_type_crucible_dsl;
 
         let gen = disk.runtime().gen;
         let name = disk.name().clone();
-        let project_id = disk.project_id;
+        let project_id = disk.project_id();
 
-        let disk: Disk = Project::insert_resource(
+        let disk_model: model::Disk = Project::insert_resource(
             project_id,
             diesel::insert_into(dsl::disk)
-                .values(disk)
+                .values(disk.model().clone())
                 .on_conflict(dsl::id)
                 .do_update()
                 .set(dsl::time_modified.eq(dsl::time_modified)),
         )
-        .insert_and_get_result_async(
-            &*self.pool_connection_authorized(opctx).await?,
-        )
+        .insert_and_get_result_async(conn)
         .await
-        .map_err(|e| match e {
-            AsyncInsertError::CollectionNotFound => authz_project.not_found(),
-            AsyncInsertError::DatabaseError(e) => public_error_from_diesel(
-                e,
-                ErrorHandler::Conflict(ResourceType::Disk, name.as_str()),
-            ),
+        .map_err(|e| {
+            err.bail(match e {
+                AsyncInsertError::CollectionNotFound => {
+                    authz_project.not_found()
+                }
+                AsyncInsertError::DatabaseError(e) => public_error_from_diesel(
+                    e,
+                    ErrorHandler::Conflict(ResourceType::Disk, name.as_str()),
+                ),
+            })
         })?;
 
-        let runtime = disk.runtime();
-        bail_unless!(
-            runtime.state().state() == &api::external::DiskState::Creating,
-            "newly-created Disk has unexpected state: {:?}",
-            runtime.disk_state
-        );
-        bail_unless!(
-            runtime.gen == gen,
-            "newly-created Disk has unexpected generation: {:?}",
-            runtime.gen
-        );
+        match &disk {
+            Disk::Crucible(CrucibleDisk { disk: _, disk_type_crucible }) => {
+                diesel::insert_into(disk_type_crucible_dsl::disk_type_crucible)
+                    .values(disk_type_crucible.clone())
+                    .on_conflict(disk_type_crucible_dsl::disk_id)
+                    .do_nothing()
+                    .execute_async(conn)
```

Comment:
hm, is `ON CONFLICT DO NOTHING` actually the correct behavior here? if we are going to go create a disk, and we discover that there are preexisting records in the `disk_type_crucible` table with the ID of the new disk we're trying to create, we probably shouldn't just leave that record there, right? that seems like it would be a bug, since those records would have old, unrelated state. OTTOMH, i can't immediately come up with a situation where that would happen except for generating a UUID that collides with one we've already used for a previous, no longer extant disk, and that does seem like it shouldn't happen. but this behavior feels like we'd be doing basically the worst possible thing in the face of something that shouldn't happen...

Reply:
This function is called by a saga node (in the disk create saga), and as a result has to be idempotent for the same input - I'm less worried about UUID collision (though in the limit we may see at least one very interesting bug as a result of this haha) and more what happens if the saga node is rerun.

I'm also generally wary of `ON CONFLICT DO NOTHING`, especially since usually the `ON CONFLICT` statement only includes the id of the row and not the other data. I'm not sure what to do instead except a `SELECT`followed by a comparison of the returned model and the argument to this function, but I think that can be elided due to the fact that UUIDs are assumed to be unique...